### PR TITLE
[PD-2547] Forms Submit

### PR DESCRIPTION
**Pull request by darrint**
This change improves the ability to submit code and handle the resulting errors.

## Test

Create contacts and contact records with problems. You should be able to navigate back to contacts and contact records and see the errors.

## Review

I've reworked the `convert` api call a bit. The overall structure is the same but the class for holding and sending errors back to the client is completely replaced.

In the client the code is much more intentional about converting `process.record` into the shape required by the API and back into the shape required by the forms.

### CODE REVIEW

File: gulp/src/nonuseroutreach/actions/process-outreach-actions.js [hunk 1]
```diff
@@ -160,53 +159,109 @@ export function extractErrorObject(fieldArray) {
...
-        const errors = {
-          partner: extractErrorObject(get(formErrors, 'partner', [])),
-          contacts: map(get(formErrors, 'contacts', []), ce =>
-            extractErrorObject(ce)),
-          communicationrecord: extractErrorObject(
-            get(formErrors, 'contactrecord', [])),
-        };
-        dispatch(noteErrorsAction(errors));
+        const apiErrors = e.data.forms;
+        if (apiErrors) {
```
Comment (tdphillips) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: What is the difference between these two ifs?

File: universal/api_validation.py [hunk 2]
```diff
@@ -92,96 +93,198 @@ def build_error_response(self):
...
+        'name': {'value': 'Alice',
+    }
+
+    to:
+
+    {'name': 'Alice'}
+    """
+    result = {}
+    for (k, v) in data.iteritems():
+        if isinstance(v, dict) and 'value' in data[k]:
```
Comment (tdphillips) [inline universal/api_validation.py]: If v is a dict, couldn't this be `'value' in v`

File: universal/api_validation.py [hunk 3]
```diff
@@ -92,96 +93,198 @@ def build_error_response(self):
...
+            result[k] = v['value']
+        elif isinstance(v, list):
+            result[k] = [collapse_values(inner) for inner in v]
+        else:
+            result[k] = collapse_values(v)
```
Comment (tdphillips) [inline universal/api_validation.py]: If a thing is not a dict or list, isn't this an infinite loop?

Comment (tdphillips) [inline universal/api_validation.py]: What type could v be if not a dict or list?

File: gulp/src/nonuseroutreach/actions/process-outreach-actions.js (hunk 1 above)
Comment (darrint) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: One is checking truthiness of `e.data`, the other of `e.data.forms`. It won't happen today but in the future `e.data` could contain something else entirely.

File: universal/api_validation.py (hunk 3 above)
Comment (darrint) [inline universal/api_validation.py]: String or int.

File: gulp/src/nonuseroutreach/actions/process-outreach-actions.js (hunk 1 above)
Comment (tdphillips) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: Both are checking e.data.forms.

File: universal/api_validation.py (hunk 3 above)
Comment (tdphillips) [inline universal/api_validation.py]: Won't that result in a 500? Neither string nor int have .iteritems

File: gulp/src/nonuseroutreach/actions/process-outreach-actions.js (hunk 1 above)
Comment (darrint) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: Good Catch...

Comment (darrint) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: Fixed.

**Comment by koonsies** [comment]
Communication type triggers a validation error when submitting the original partner form, which causes the communication record card to appear. Beyond that, the error for not having a communication record selected is displaying as "Value u'{}' is not a valid choice."

This also causes issues with the workflow PR I put up before. Since that CommRecord object is present after validation on partner, it tries to skip the comm record step.

**Comment by koonsies** [comment]
It's not directly related to this PR, but I noticed that ContactRecord's datetime field has a blank=True set, but not null=True. This is causing the full_clean method to say that having a blank date time is fine, but when it goes to save the model, it is triggering a 500 error. We need to look into why this field is set up this way (without having a default value) in the future.

**Comment by tdphillips** [comment]
There is no meaningful default. Any default value we could pick for "date this event happened" would almost certainly be wrong.

**Pull request merged**